### PR TITLE
feat: Refactor Helm chart

### PR DESCRIPTION
**Pull request by stevehipwell**
**1. Issue, if available:**
This PR addresses part of #781.

**2. Description of changes:**

This is a breaking change.

- Improve chart metadata
- Make chart follow idiomatic Helm patterns
- Use K8s recommended labelling to distinguish between components and improve selector logic
- Use a different ServiceAccount for controller and webhook, this is more secure
- Recommend creating service accounts rather than using the namespace default, this is more secure

**3. How was this change tested?**

The template output was compared to the pervious version.

**4. Does this change impact docs?**
- [x] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by stevehipwell** [comment]
@ellistarn sorry it took me so long to come through on this one, but with the holidays and the chart refactor I did for NTH taking longer than expected I've got quite a backlog of pending OSS commitments. I'm about to go through the issues and comment on the ones that this PR is related to.

File: charts/karpenter/Chart.yaml
Comment (mbevc1) [inline charts/karpenter/Chart.yaml:10]: Probably a typo and we want something like autoscaling and karpenter here :thinking:

Comment (stevehipwell) [reply charts/karpenter/Chart.yaml:10]: @mbevc1 I completely forgot to put in the correct keywords. These are from the chart I used as a template. My suggestion would be as follows.

- cluster
- node
- scheduler
- autoscaling
- lifecycle

**Comment by stevehipwell** [comment]
@ellistarn have you had a chance to review this yet?

**Comment by bwagner5** [comment]
I'll take a look at this today @stevehipwell !  Looks familiar ;)

**Review by bwagner5 — COMMENTED** [review]
I haven't quite finished reviewing all the files, but had a question on the serviceAccount

File: Makefile
Comment (stevehipwell) [reply Makefile:74]: Shouldn't the annotation be passed in with `--set` flag? Otherwise you're using the namespace default Service Account and giving it all the access. I guess for development it doesn't really matter, but this pattern was also copied everywhere as is often the case.

**Comment by ellistarn** [comment]
FYI related: https://github.com/aws/karpenter/pull/914/files

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (bwagner5) [inline charts/karpenter/templates/controller/serviceaccount.yaml:1]: since we now have a controller service account and a webhook service account, the IAM Roles for Service Accounts docs would need to be updated to create both or instructions to annotate both. I'm not sure breaking these apart is worth the extra setup work. 

I actually think we could just combine the controller and webhook containers into the same pod. There would be multiple services for the pod but that's doable with the named ports. This would simplify a lot of the redundant parameters between the webhook and controller in the values.yaml. WDYT?

File: Makefile
Comment (bwagner5) [reply Makefile:74]: We thought users would be using the getting started guide eksctl command to create the IAM Role for Service Accounts, which creates the service account and annotates with the role arn (https://karpenter.sh/v0.5.5/getting-started/#create-the-karpentercontroller-iam-role).   That might be a bad assumption for production installs, I'm not sure. But for dev installs, it's probably accurate and so leaving the `--set serviceAccount.controller.create=false` seems like the best dev workflow experience to me. Otherwise, we'd need some way of persisting the role arn in an env var or something and then using a set to annotate every time we `make apply`.

Comment (stevehipwell) [reply Makefile:74]: That guide really needs re-writing to promote better practice, although from memory one of the other tabs has a better way of creating the roles and getting the ARN back? It's idiomatic for a Helm chart to create it's own service account(s), so I'd suggest documenting the externally created one, but with a name specified so it's not the default namespace one.

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (stevehipwell) [reply charts/karpenter/templates/controller/serviceaccount.yaml:5]: I did, it's idiomatic to only set the namespace in templates where you're overriding the value provided to the `--namespace` flag.

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (stevehipwell) [reply charts/karpenter/templates/controller/serviceaccount.yaml:1]: The K8s roles were already separate, so it would be correct to split the IAM roles; but duplicating the IRSA annotation would make it behave as it currently does. Merging the pods would probably be best in the long run.

File: Makefile
Comment (stevehipwell) [reply Makefile:74]: For the dev workflow, once the role has been created it persists outside of the K8s resources. It'd be simplest to set it in the env.

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (bwagner5) [reply charts/karpenter/templates/controller/serviceaccount.yaml:5]: interesting, I did not realize this! :)

File: Makefile
Comment (bwagner5) [reply Makefile:74]: Yeah, looks like the terraform getting started guide does the annotation by passing to helm: https://karpenter.sh/v0.5.5/getting-started-with-terraform/#install-karpenter-helm-chart 

Speaking of which, it looks like that doc needs to be updated with this change since the `controller.clusterName` is now at the root level.

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (bwagner5) [reply charts/karpenter/templates/controller/serviceaccount.yaml:1]: Did you want to take on the merging of the webhook and controller to one pod in order to simplify the helm chart?

File: Makefile
Comment (bwagner5) [reply Makefile:74]: also, agreed on the annotating. Can you update the docs and Makefile to orchestrate that process?

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (stevehipwell) [reply charts/karpenter/templates/controller/serviceaccount.yaml:1]: Could we leave that until this has been merged as it's adding another level of complexity or do you want all of the pain now?

Comment (bwagner5) [reply charts/karpenter/templates/controller/serviceaccount.yaml:1]: I think it would simplify the helm chart quite a bit (getting rid of the controller and webhook block, removing service account duplication, etc), so I'd advocate for all the pain now :D

Comment (stevehipwell) [reply charts/karpenter/templates/controller/serviceaccount.yaml:1]: OK, I'll do it.

Comment (stevehipwell) [reply charts/karpenter/templates/controller/serviceaccount.yaml]: @bwagner5 I've just done this, and the chart does look a lot cleaner.

File: Makefile
Comment (stevehipwell) [reply Makefile:76]: @bwagner5 does the instance profile change make a difference to any of this? Secondly which docs need updating?

In a related side discussion I think the chart needs decoupling from the binary release so I'll open a new issue for this.

Comment (stevehipwell) [reply Makefile:76]: FYI https://github.com/aws/karpenter/issues/1253

File: website/content/en/v0.5.5/getting-started/_index.md
Comment (bwagner5) [inline website/content/en/v0.5.5/getting-started/_index.md:145]: IIUC this should include the IRSA annotations if we remove this, or else it will override the service account created above with eksctl?

File: Makefile
Comment (bwagner5) [reply Makefile:76]: The instance profile shouldn't make a difference since it's for the nodes that karpenter provisions and not the karpenter controller instance role.

File: charts/karpenter/templates/controller/serviceaccount.yaml
Comment (bwagner5) [reply charts/karpenter/templates/controller/serviceaccount.yaml]: nice! so much cleaner! Thanks for doing that!!

File: Makefile
Comment (bwagner5) [reply Makefile:76]: The docs that need updating are here (I think that's all of the places):

 - https://karpenter.sh/v0.6.0/getting-started/#install-karpenter-helm-chart (may want to move the helm install docs above the spot service linked role docs since we need to pass thru the annotation now. Probably want to add `--role-only` to the eksctl cmd to prevent creating the service account too)
 - https://karpenter.sh/v0.6.0/getting-started-with-terraform/#install-karpenter-helm-chart 
 - https://karpenter.sh/v0.6.0/development-guide/#environment-specific-setup

Comment (stevehipwell) [reply Makefile:76]: @bwagner5 shouldn't we be waiting until there is a release to make the changes to the docs for that release? The `v0.6.0` chart has already been released so I don't think the docs shouldn't change for that version? This is one of the problems of directly linking your chart releases to your package releases.

While I wait for a response I'll update the non-docs references and undo the change to docs I accidentally made previously.

Comment (stevehipwell) [reply Makefile:61]: @bwagner5 @ellistarn how do the docs get updated for a release?

Comment (bwagner5) [reply Makefile:61]: All docs before a release go in the preview directory https://github.com/aws/karpenter/tree/main/website/content/en/preview 

When we cut a release, the preview directory is copied to the version we are releasing, so it's fine to make the doc updates to preview in this PR.

Comment (stevehipwell) [reply Makefile:61]: These should be updated now.

File: website/content/en/preview/getting-started/_index.md
Comment (bwagner5) [inline website/content/en/preview/getting-started/_index.md:312]:
```suggestion
aws iam delete-role --role-name "${CLUSTER_NAME}-karpenter"
```

**Review by bwagner5 — CHANGES_REQUESTED** [review]
Nice! Just one comment on a mistyped command in the docs

**Comment by stevehipwell** [comment]
@bwagner5 thanks for the review, I've fixed the typo you found. Hopefully this can be merged soon.

**Comment by stevehipwell** [comment]
@bwagner5 the CI is failing but I don't think it's related to any changes I made?

**Comment by bwagner5** [comment]
I pulled this down and did some manual testing. I posted a PR to your fork with a few changes: https://github.com/stevehipwell/karpenter/pull/1 

We're very near a merge! I love the single karpenter pod!

**Comment by stevehipwell** [comment]
@bwagner5 thanks for the review and the PR.

**Comment by mbevc1** [comment]
Nice work and also glad to see consolidating Pod too :+1:

**Comment by stevehipwell** [comment]
@bwagner5 all merged and in this PR.

**Review by bwagner5 — APPROVED** [review]
LGTM! Thanks!!

**Comment by Skarlso** [comment]
In the future please note that refactors like these break `eksctl`'s Karpenter installation flow. A heads up would have been nice.

**Comment by metanerd** [comment]
Hi! I just want to mention that this refactor also broke the values for this aws sample: https://github.com/aws-samples/karpenter-terraform/blob/main/karpenter/templates/values.yaml.tpl

**Comment by stevehipwell** [comment]
@metanerd I'm just a contributor so just my opinion, but all samples _SHOULD_ use fixed versions and all samples **MUST** fix the version of any pre v1.0.0 components.

**Pull request merged**
